src/theory/uf/proof_checker.cpp

```diff
@@ -202,6 +203,29 @@ Node UfProofRuleChecker::checkInternal(PfRule id,
     Node ret = TheoryUfRewriter::getHoApplyForApplyUf(args[0]);
     return args[0].eqNode(ret);
   }
+  else if (id == PfRule::BETA_REDUCE)
+  {
+    Assert(args.size() >= 2);
+    Node lambda = args[0];
+    if (lambda.getKind() != LAMBDA)
+    {
+      return Node::null();
+    }
+    std::vector<TNode> vars(lambda[0].begin(), lambda[0].end());
+    std::vector<TNode> subs(args.begin() + 1, args.end());
+    if (vars.size() != subs.size())
+    {
+      return Node::null();
+    }
+    NodeManager* nm = NodeManager::currentNM();
+    std::vector<Node> appArgs;
+    appArgs.push_back(lambda);
+    appArgs.insert(appArgs.end(), subs.begin(), subs.end());
+    Node app = nm->mkNode(APPLY_UF, appArgs);
+    Node ret = lambda[1].substitute(
```

Comment:
Shouldn't we maybe use `substituteCaptureAvoiding` here? There are instances in the solver where that is actually used and this checking would break here. (In that case the documentation should be updated to say the substitution is applied in a capture avoiding manner). We'd need to do the canonical bound var stuff, though...

Reply:
Capture-avoiding substitution is deprecated, I am planning to eliminate that method.  We now handle introduced variable shadowing in the quantifiers rewriter.

We currently still rely on the capture-avoiding substitution for beta reduction in the UF rewriter, however reconstruction to BETA_REDUCE will fail gracefully if there is a discrepancy.

Reply:
Sounds good! In that case I'd just like to have in the documentation of the rule that its semantics is beta reduction with regular substitution.